server/src/main/java/org/elasticsearch/index/mapper/SourceFieldMapper.java

```diff
@@ -324,7 +360,7 @@ public SourceFieldMapper build() {
             }
 
             SourceFieldMapper sourceFieldMapper = new SourceFieldMapper(
-                sourceMode,
+                INDEX_MAPPER_SOURCE_MODE_SETTING.exists(settings) ? INDEX_MAPPER_SOURCE_MODE_SETTING.get(settings) : mode.get(),
```

Comment:
We need to to this because the `if(isDefault())` does not kick in in some cases (non-default cases).